src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs

```diff
@@ -0,0 +1,367 @@
+        /// <param name="kind">The <see cref="ActivityKind"/></param>
+        /// <param name="parentContext">The parent <see cref="ActivityContext"/> object to initialize the created Activity object with.</param>
+        /// <param name="tags">The optional tags list to initialize the created Activity object with.</param>
+        /// <param name="links">The optional <see cref="ActivityLink"/> list to initialize the created Activity object with.</param>
+        /// <param name="startTime">The optional start timestamp to set on the created Activity object.</param>
+        /// <returns>The created <see cref="Activity"/> object or null if there is no any listener.</returns>
+        public Activity? StartActivity(string name, ActivityKind kind, ActivityContext parentContext, IEnumerable<KeyValuePair<string, string?>>? tags = null, IEnumerable<ActivityLink>? links = null, DateTimeOffset startTime = default)
+            => StartActivity(name, kind, parentContext, null, tags, links, startTime);
+
+        /// <summary>
+        /// Creates a new <see cref="Activity"/> object if there is any listener to the Activity events, returns null otherwise.
+        /// </summary>
+        /// <param name="name">The operation name of the Activity.</param>
+        /// <param name="kind">The <see cref="ActivityKind"/></param>
+        /// <param name="parentId">The parent Id to initialize the created Activity object with.</param>
+        /// <param name="tags">The optional tags list to initialize the created Activity object with.</param>
+        /// <param name="links">The optional <see cref="ActivityLink"/> list to initialize the created Activity object with.</param>
+        /// <param name="startTime">The optional start timestamp to set on the created Activity object.</param>
+        /// <returns>The created <see cref="Activity"/> object or null if there is no any listener.</returns>
+        public Activity? StartActivity(string name, ActivityKind kind, string parentId, IEnumerable<KeyValuePair<string, string?>>? tags = null, IEnumerable<ActivityLink>? links = null, DateTimeOffset startTime = default)
+            => StartActivity(name, kind, default, parentId, tags, links, startTime);
+
+        private Activity? StartActivity(string name, ActivityKind kind, ActivityContext context, string? parentId, IEnumerable<KeyValuePair<string, string?>>? tags, IEnumerable<ActivityLink>? links, DateTimeOffset startTime)
+        {
+            // _listeners can get assigned to null in Dispose.
+            SynchronizedList<ActivityListener>? listeners = _listeners;
+            if (listeners == null || listeners.Count == 0)
+            {
+                return null;
+            }
+
+            Activity? activity = null;
+
+            ActivityDataRequest dateRequest = ActivityDataRequest.None;
+
+            if (parentId != null)
+            {
+                listeners.EnumWithFunc(listener => {
```

Comment:
Performance: We should benchmark, but I am suspicious the synchronization and versioning may add enough overhead that the pre-existing DiagnosticSource pattern will wind up being faster. If benchmarking confirmed the overhead here was meaningful we could optimize it:

Given that adding/removing listeners from a source listener list will be a rare action and the list will be small it is probably reasonable to create a new immutable array each time the list is modified and assigning that array to the ActivitySource._listeners variable represents an atomic transition from the old list to the new list. When dispatching activity events to the listeners the source could then capture the _listeners field into a local variable so that it has an immutable snapshot of the list to iterate over. This should eliminate the need for lock acquisition as well as any version checking.

Reply:
Yes, we can do the benchmarking later and optimize more as needed. Thanks.